datafusion/common/src/tree_node.rs

```diff
@@ -368,7 +371,7 @@ impl<T: DynTreeNode + ?Sized> TreeNode for Arc<T> {
             let arc_self = Arc::clone(&self);
             self.with_new_arc_children(arc_self, new_children?)
         } else {
-            Ok(self)
+            Ok(self.clone())
```

Comment:
Just one minor note, that this clone seems to be unnecessary.

Reply:
Yea, probably I added it during trying different ideas locally. Removed.